### DISCUSSION TITLE
NULL errors are not managed by the comparison levels.

### DISCUSSION THREAD

**Discussion by jcamatta** [moj-analytical-services/splink#2494] [Q&A] [answered]
Hello, how are you?

Could you help me with this problem?, please

I have a model_settings which has a comparison rule with a null level like the next one:
```json
{
    "output_column_name": "PHONE",
    "comparison_levels": [
        {
            "sql_condition": "\"PHONE_l\" IS NULL OR \"PHONE_r\" IS NULL",
            "label_for_charts": "PHONE is NULL",
            "is_null_level": true
        },
        {
            "sql_condition": "jaro_similarity(\"PHONE_l\", \"PHONE_r\") >= 0.9",
            "label_for_charts": "Jaro distance of 'PHONE >= 0.9'",
            "m_probability": 0.026548979480589187,
            "u_probability": 1.0423679520484444e-05
        }
    ]
}
```

But if the data I pass to the model has ALL values in the column as None this throw me an error when executing `linker.inference.predict`:

```python
BinderException: Binder Error: No function matches the given name and argument types 'jaro_similarity(INTEGER, "NULL")'. You might need to add explicit type casts.
	Candidate functions:
	jaro_similarity(VARCHAR, VARCHAR) -> DOUBLE
```
This is the data test case I am using:
![image](https://github.com/user-attachments/assets/b7882b5f-8c5a-4978-8a8c-c52dba57ea60)

I don't know if this is an issue or is normal behaviour. I was expecting that the Null level to be as a filter, so the next rules doesn't execute if at least one of the value in comparison is null. I mean, it has sense that I don't need to checkout if the similarity is >= 0.7 if has already passed the >= 0.9.

In the same exception I see the next SQL which is pointed in the direction I just said:
```sql
CASE
      WHEN "PHONE_l" IS NULL OR "PHONE_r" IS NULL
      THEN -1
      WHEN "PHONE_l" = "PHONE_r"
      THEN 3
      WHEN JARO_SIMILARITY("PHONE_l", "PHONE_r") >= 0.9
      THEN 2
      WHEN JARO_SIMILARITY("PHONE_l", "PHONE_r") >= 0.8
      THEN 1
      ELSE 0
    END AS gamma_PHONE,
```

Note that for the gender column I don't have that problem, because the rule I use is the ExactMatch.

**Comment by jcamatta** [comment]
The only temporal fix I found is to create a random person that doesn't match with anyone but has all properties as not nulls.

**Reply by RobinL** [reply, marked as answer]
The problem here is that the backend database (duckdb) incorrectly infers the data type of the column.  If there are no values in it, it has no basis to guess the data type.

Instead, you need to create a table with the desired schema (phone being a string here presumably), and pass data into Splink as duckdb tables (duckdbpyrelation) as opposed to going through pandas.

So long as the data going into Splink has the correct schema for the similarity functions, you won't get the error

**Reply by jcamatta** [reply]
Oh, I didn't know that. Thanks you!
